File: src/main/java/org/kite9/diagram/visualization/orthogonalization/flow/EdgeVertex.java
package org.kite9.diagram.visualization.orthogonalization.flow;

import java.util.Set;

import org.kite9.diagram.common.elements.edge.Edge;
import org.kite9.diagram.common.elements.vertex.AbstractVertex;
import org.kite9.diagram.common.elements.vertex.Vertex;
import org.kite9.diagram.model.DiagramElement;

public class EdgeVertex {

	private static Vertex NULL_VERTEX =  new AbstractVertex("NULL") {

		@Override
		public Set<DiagramElement> getDiagramElements() {
			return null;
		}

		@Override
		public boolean isPartOf(DiagramElement de) {
			return false;
		}
	};
	
	Edge edge;
	Vertex vertex = NULL_VERTEX;

	@Override
	public boolean equals(Object arg0) {
		if (arg0 instanceof EdgeVertex) {
			EdgeVertex ff = (EdgeVertex) arg0;
			return ff.edge.equals(edge) && ff.vertex.equals(vertex);
		}
		return false;
	}

	@Override
	public int hashCode() {
		return edge.hashCode() + vertex.hashCode();
	}

	public String toString() {
		return edge.toString() + "-" + vertex.getID();
	}
}

File: src/main/kotlin/org/kite9/diagram/visualization/orthogonalization/flow/EdgeVertex.kt
package org.kite9.diagram.visualization.orthogonalization.flow

import org.kite9.diagram.common.elements.edge.Edge
import org.kite9.diagram.common.elements.vertex.AbstractVertex
import org.kite9.diagram.common.elements.vertex.Vertex
import org.kite9.diagram.model.DiagramElement

data class EdgeVertex(val edge: Edge, val vertex: Vertex) {

    constructor(edge: Edge) : this(edge, NULL_VERTEX)

    companion object {
        private val NULL_VERTEX: Vertex = object : AbstractVertex("NULL") {
            override fun getDiagramElements(): Set<DiagramElement> {
                return emptySet()
            }

            override fun isPartOf(de: DiagramElement?): Boolean {
                return false
            }
        }
    }
}